### PR TITLE
variable substitution for azure rm web app

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+function Get-MimeType()
+{
+    param($extension = $null);
+ 
+    $mimeType = $null;
+    
+    if ( $null -ne $extension )
+    {
+        $drive = Get-PSDrive HKCR -ErrorAction SilentlyContinue;
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: Does this require Administrator privilege to look up? The Task process may not have Admin rights. Can you check.

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    $mimeType;
+}
+
+
+# getting variables defined in environment
+$definedVariables = Get-ChildItem Env:
+
+# Find all configuration files in extracted path of webapp
+$webConfigFiles = Get-ChildItem -Path $WebAppFolderPath -Filter $ConfigFileRegex -Recurse -ErrorAction SilentlyContinue -Force
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: If you don't find any files, you would want to log and return.

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+foreach( $file in $WebConfigFiles )
+{
+    # Get mime type of each file
+    $fileMimeType = Get-MimeType -extension $file.Extension
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: If you don't find the right extension, you would want to log that you cannot detect the file type.

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    foreach($variable in $definedVariables){
+
+        # Compare mime type of file and invoke corresponding variable substituter
+        if( $fileMimeType -eq $ApplicationXMLMimeType ){
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: As said above, I am not sure if you will get exact match, if the Content Type value differs in different PCs. You might want to just restrict to finding substring xml. Thoughts?

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+            .\XMLSubstituter.ps1 -xmlFile $file.FullName -tag "appSettings" -conditionalAttributeKey "key" -conditionalAttributeValue "$($variable.Name)" -targetKey "value" -targetValue "$($variable.Value)"
+            .\XMLSubstituter.ps1 -xmlFile $file.FullName -tag "MyConnectionString"  -targetKey "$($variable.Name)" -targetValue "$($variable.Value)"
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: For both appsettings and connection string, the value can have special characters. Do you have to escape them in same way you did it for few other tasks?

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 6]
```diff
@@ -0,0 +1,74 @@
+﻿Param(
+  [Parameter(mandatory=$true)][string]$xmlFilePath,
+  [Parameter(mandatory=$true)][string]$tag,
+  [string]$conditionalAttributeKey,
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: For consistency, can you add the Parameter and mandatory attributes (with false) here.

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  [string]$conditionalAttributeValue,
+  [Parameter(mandatory=$true)][string]$targetKey,
+  [Parameter(mandatory=$true)][string]$targetValue
+)
+
+Write-Verbose "xmlFile = $xmlFilePath" -Verbose
+Write-Verbose "Tag = $tag" -Verbose
+Write-Verbose "conditionalAttribute = $conditionalAttributeKey" -Verbose
+Write-Verbose "conditionalValue = $conditionalAttributeValue" -Verbose
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: Can this have secrets? If so, we should not print them.

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+Write-Verbose "targetKey = $targetKey" -Verbose
+Write-Verbose "targetValue = $targetValue" -Verbose
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: Can this have secrets? If so, we should not print them.

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+<#
+   UpdateXMLAttribute function takes xml node as argument and check if for any node it match with 
+   target key for which value need to be changed, It update attribute value. Otherwise, it recursivly
+   checks for all its child node.
+#>
+function UpdateXMLAttribute
+{
+
+    Param(
+          [Parameter(mandatory=$true)][string]$tag,[string] $conditionalAttributeKey , [string] $conditionalAttributeValue ,[string] $targetKey, [string] $targetValue ,[Xml.XmlElement] $xml
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: For readability, can you have each parameter in separate line.

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                if(  (([string]::IsNullOrEmpty($conditionalAttributeKey) -eq $false -and  $childNode.GetAttribute($conditionalAttributeKey) -eq $conditionalAttributeValue) -or ([string]::IsNullOrEmpty($conditionalAttributeKey))) -and $childNode.GetAttribute($targetKey) -ne $null){
+                    $childNode.SetAttribute($targetKey,$targetValue)
+                    return
+                }
+            }
+            
+        }
+
+        # If node has child recursivly check for each of child for properties
+        if($node.ChildNodes.Count -gt 0 )
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: Can we sync-up on this recursion part. Have few questions.

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 (hunk 1 above)
Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: No it doesn't require admin rights

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 (hunk 4 above)
Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: There are many MIME type with same substring and we might not be interested to look into those MIME types. eg. application/vnd.las.las+xml

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 (hunk 6 above)
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: inordewr to be consistent with syntax, use [Parameter(mandatory=$false)] and also first specify object type and then whether it is mandatory or not . like [String][Parameter(Mandatory=$true)] $webAppName check checked in task in master branch

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 11]
```diff
@@ -0,0 +1,74 @@
+﻿Param(
+  [Parameter(mandatory=$true)][string]$xmlFilePath,
+  [Parameter(mandatory=$true)][string]$tag,
+  [string]$conditionalAttributeKey,
+  [string]$conditionalAttributeValue,
+  [Parameter(mandatory=$true)][string]$targetKey,
+  [Parameter(mandatory=$true)][string]$targetValue
+)
+
+Write-Verbose "xmlFile = $xmlFilePath" -Verbose
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: -verbose is not recommended now. if we set system.debug true all these statements are printed without -verbose verb and if you want to specify in console window then use write-host get-localized...refer checked in task in master

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+Write-Verbose "Tag = $tag" -Verbose
+Write-Verbose "conditionalAttribute = $conditionalAttributeKey" -Verbose
+Write-Verbose "conditionalValue = $conditionalAttributeValue" -Verbose
+Write-Verbose "targetKey = $targetKey" -Verbose
+Write-Verbose "targetValue = $targetValue" -Verbose
+
+
+if( $conditionalAttribute -eq $null -xor $conditionalValue -eq $null) {
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: use String::IsNullOrEmpty($param) which is recommended in powershell for null checks

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+  [Parameter(mandatory=$true)][string]$webAppFolderPath,
+  [Parameter(mandatory=$true)][string]$ConfigFileRegex
+)
+
+# define mime supported mime type
+$ApplicationXMLMimeType = "application/xml" 
+
+function Get-MimeType()
+{
+    param($extension = $null);
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: you can remove semicolons. Not required in powershell. applicable every where

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+ 
+    $mimeType = $null;
+    
+    if ( $null -ne $extension )
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: Use -not String::IsNullOrEmpty($param) which is recommended in powershell for null checks

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 (hunk 13 above)
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: you can use [Object][Parameter(Mandatory=$false)] here

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+            $drive = New-PSDrive -Name HKCR -PSProvider Registry -Root HKEY_CLASSES_ROOT
+        }
+        $mimeType = (Get-ItemProperty HKCR:$extension)."Content Type";
+    }
+    $mimeType;
+}
+
+
+# getting variables defined in environment
+$definedVariables = Get-ChildItem Env:
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: name it definedEnvVariables

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 (hunk 13 above)
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: name it fileExtension

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    if ( $null -ne $extension )
+    {
+        $drive = Get-PSDrive HKCR -ErrorAction SilentlyContinue;
+        if ( $null -eq $drive )
+        {
+            $drive = New-PSDrive -Name HKCR -PSProvider Registry -Root HKEY_CLASSES_ROOT
+        }
+        $mimeType = (Get-ItemProperty HKCR:$extension)."Content Type";
+    }
+    $mimeType;
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: use return $mimeType here

File: Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1 [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    $mimeType = $null;
+    
+    if ( $null -ne $extension )
+    {
+        $drive = Get-PSDrive HKCR -ErrorAction SilentlyContinue;
+        if ( $null -eq $drive )
+        {
+            $drive = New-PSDrive -Name HKCR -PSProvider Registry -Root HKEY_CLASSES_ROOT
+        }
+        $mimeType = (Get-ItemProperty HKCR:$extension)."Content Type";
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/VariableSustituter.ps1]: You can add a verbose log found mimetype "" for filextension as : ""

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    Write-Verbose "Either conditionalAttribute and conditionalValue should be set or both should be blank" -Verbose
+}
+
+if (-not (test-path $xmlFilePath)) {
+    throw "XML file path is not present."
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: use Get-LocalizedString Syntax here applicable everywhere

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+        return
+    }
+
+    foreach( $node in $xml.ChildNodes )
+    {
+        
+        if( $node.Name -eq $tag ) 
+        {
+            foreach( $childNode in $node.ChildNodes ){   
+                if(  (([string]::IsNullOrEmpty($conditionalAttributeKey) -eq $false -and  $childNode.GetAttribute($conditionalAttributeKey) -eq $conditionalAttributeValue) -or ([string]::IsNullOrEmpty($conditionalAttributeKey))) -and $childNode.GetAttribute($targetKey) -ne $null){
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: Use -not [string]::IsNullOrEmpty($conditionalAttributeKey) and it looks like we are computing things and comparing in same if clause can you evaluate then before if clause and then have a check in if clause

File: Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1 [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+        # If node has child recursivly check for each of child for properties
+        if($node.ChildNodes.Count -gt 0 )
+        {
+           UpdateXMLAttribute -xml $node -tag $tag -conditionalAttributeKey $conditionalAttributeKey -conditionalAttributeValue $conditionalAttributeValue -targetKey $targetKey -targetValue $targetValue
+        }
+    }
+}
+
+
+$xml = [xml](Get-Content $xmlFilePath)
```
Comment (anaggar) [inline Tasks/AzureRmWebAppDeployment/XMLSubstituter.ps1]: name in xmlFileContent

**Comment by anaggar** [comment]
please add pester tests for these files

**Pull request closed without merging**